apps/tools/blocks-crawler/src/blocks.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2020 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0
+
+use mina_rs_base::types::ExternalTransition;
+use mina_rs_base::JsonSerializationType;
+use serde::{Deserialize, Serialize};
+use serde_json::Error;
+
+// To read query response from graphql api (https://graphql.minaexplorer.com/)
+#[derive(Debug, Serialize, Deserialize)]
+pub struct QueryResponse {
+    pub data: Blocks,
+}
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Blocks {
+    pub blocks: Vec<BlockInfo>,
+}
+#[derive(Debug, Serialize, Deserialize)]
+pub struct BlockInfo {
+    #[serde(rename = "blockHeight")]
+    pub block_height: usize,
+    #[serde(rename = "stateHash")]
+    pub state_hash: String,
+}
+
+impl std::fmt::Display for BlockInfo {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}-{}", self.block_height, self.state_hash)
+    }
+}
+
+pub async fn check_block(block: BlockInfo) {
+    info!("Checking: {}", block);
+    // Retrieve block json from mainnet json dump
+    match retrieve_block_json("mainnet", &block).await {
+        Ok(block_json) => match parse_block_json(&block_json) {
+            Ok(_et) => {
+                write_json("succeeded", "mainnet", &block, &block_json);
+            }
+            Err(err) => {
+                write_json("failed", "mainnet", &block, &block_json);
+                error!(
+                    "Fix Me(https://storage.googleapis.com/mina_network_block_data/mainnet-{}.json): {}",
+                    block, err
+                );
+            }
+        },
+        Err(err) => {
+            warn!("Failed to retrieve block json: {}", err);
+        }
+    };
+}
+
+pub async fn fetch_block_info(n: &str) -> Result<QueryResponse, reqwest::Error> {
+    let client = reqwest::Client::new();
+    let res = client
+        .post("https://graphql.minaexplorer.com/")
+        // Below query retrieves blockHeight and stateHash of 10 latest blocks
+        // To learn more about graphql query refer to https://docs.minaprotocol.com/en/developers/graphql-api 
+        .json(&serde_json::json!({
+            "query":"query Fetch {\n  blocks(limit: $n, sortBy: RECEIVEDTIME_DESC) {\n    blockHeight\n    stateHash\n  }\n}".replace("$n", n)
```

Comment:
Maybe use `format!(r###"
  formatted json
"###)` for better readability

Reply:
And better to use string formatting instead of replace

Reply:
fixed